src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobTask.cs

```diff
@@ -99,28 +156,147 @@ public async Task ExecuteAsync(ReindexJobRecord reindexJobRecord, WeakETag weakE
             }
             catch (Exception ex)
             {
-                // The job has encountered an error it cannot recover from.
-                // Try to update the job to failed state.
-                _logger.LogError(ex, "Encountered an unhandled exception. The job will be marked as failed.");
-
                 _reindexJobRecord.Error.Add(new OperationOutcomeIssue(
                     OperationOutcomeConstants.IssueSeverity.Error,
                     OperationOutcomeConstants.IssueType.Exception,
                     ex.Message));
-                await CompleteJobAsync(OperationStatus.Failed, cancellationToken);
+
+                _reindexJobRecord.FailureCount++;
+
+                _logger.LogError(ex, $"Encountered an unhandled exception. The job failure count increased to {_reindexJobRecord.FailureCount}.");
+
+                await UpdateJobAsync(cancellationToken);
+
+                if (_reindexJobRecord.FailureCount >= _reindexJobConfiguration.ConsecutiveFailuresThreshold)
+                {
+                    await CompleteJobAsync(OperationStatus.Failed, cancellationToken);
+                }
+            }
+        }
+
+        private async Task ProcessQueryAsync(ReindexJobQueryStatus query, CancellationToken cancellationToken)
+        {
+            try
+            {
+                query.Status = OperationStatus.Running;
+                query.LastModified = DateTimeOffset.UtcNow;
+
+                // Query first batch of resources
+                var results = await ExecuteReindexQueryAsync(query, false, cancellationToken);
+
+                // if continuation token then update next query
+                if (!string.IsNullOrEmpty(results.ContinuationToken))
+                {
+                    var nextQuery = new ReindexJobQueryStatus(results.ContinuationToken);
+                    nextQuery.LastModified = DateTimeOffset.UtcNow;
+                    nextQuery.Status = OperationStatus.Queued;
+                    _reindexJobRecord.QueryList.Add(nextQuery);
+                }
+
+                await UpdateJobAsync(cancellationToken);
+
+                // TODO: Release lock on job document so another thread may pick up the next query.
+
+                await _updateIndices.ProcessSearchResultsAsync(results, _reindexJobRecord.Hash, cancellationToken);
```

Comment:
Should this be the new hash or the old hash here?

Reply:
This should be the new/current hash value.  I have updated the search logic to be correct.